### ISSUE TITLE
Flutter Web Semantics does not announce the same announcement multiple times in a row

### ISSUE THREAD

**Issue by tugorez** [flutter/flutter#142250]
### Steps to reproduce

Run the sample app and click the `Different announcements` and `Same announcements` buttons multiple times.

### Expected results

Similar to how all the `Different announcements` generated are being read and announced by the screen reader, all the `Same announcements` should be read/announced as well by the screen reader.

### Actual results

When "announcing" messages with the same content some of them are being dropped.

### Code sample

<details open><summary>Code sample</summary>

```dart
import 'package:flutter/semantics.dart';
import 'package:flutter/material.dart';

const Color darkBlue = Color.fromARGB(255, 18, 32, 47);

var count = 0;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark().copyWith(
        scaffoldBackgroundColor: darkBlue,
      ),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        OutlinedButton(
          onPressed: () async {
            final textDirection = Directionality.of(context);
            await SemanticsService.announce('Test announcement $count', textDirection, assertiveness: Assertiveness.assertive);
            count++;
          },
          child: Text(
            'Different announcements',
            style: Theme.of(context).textTheme.headlineMedium,
          ),
        ),
        OutlinedButton(
          onPressed: () async {
            final textDirection = Directionality.of(context);
            await SemanticsService.announce('Test announcement', textDirection,  assertiveness: Assertiveness.assertive);
          },
          child: Text(
            'Same announcements',
            style: Theme.of(context).textTheme.headlineMedium,
          ),
        ),
      ],
    );
  }
}
```

</details>

**Comment by tugorez** [comment]
https://github.com/flutter/flutter/issues/127335 and https://github.com/flutter/engine/pull/42258 have relevant context.

**Comment by tugorez** [comment]
My initial guess is that 

1. By not providing an [aria-relevant][1] value when [creating the live regions][2], we are defaulting to [additions text][3]
2. This means the user agent will be looking for node and/or text additions within the region.
3. Given that we're removing elements and adding new ones with exactly the same content, it makes the user agent believe nothing has changed.

One potential solution, that seems to be working locally, would be to add a suffix (like the one removed in https://github.com/flutter/engine/pull/42258) but with a little twist: Alternate its addition.

```dart
                                             
 bool _shouldAddMessageSuffix = false; // added                          
 static const _messageSufix = '.';  //added
 
 void announce(String message, Assertiveness assertiveness) {                  
   final DomHTMLElement ariaLiveElement = ariaLiveElementFor(assertiveness);   
   final DomHTMLDivElement messageElement = createDomHTMLDivElement();         
   final String text = _shouldAddMessageSuffix ? '$message $_messageSufix' : message; // added
   _shouldAddMessageSuffix = !_shouldAddMessageSuffix;                         
                                                                               
   messageElement.text = text; // changed
   ariaLiveElement.append(messageElement);                                     
   Timer(liveMessageDuration, () => messageElement.remove());                  
 }                                                                             
```

[1]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-relevant
[2]: https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/semantics/accessibility.dart#L101
[3]: https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-relevant#additions_text

**Comment by marcianx** [comment]
Removal and addition doesn't happen at the same time though and repeated vocalizations should work just fine.

I just tested out [the sample code in the issue on DartPad](https://dartpad.dev/?id=66ddb3f28b71cbbf30ee35102e48f1b4) using ChromeVox and both buttons worked fine and repeat announcements, even when I rapidly activated the "Same announcements" button. Also, these announcements are polite -- so screen readers are totally allowed to elide some of them if they are coming in rapid succession. You can always [pass in `Assertiveness.assertive` to `SemanticsService.announce`](https://api.flutter.dev/flutter/semantics/SemanticsService/announce.html) to see if it avoid dropping stuff, but really dropping announcements should only happen for polite if there's a lot of competing vocalization going on or if the tab doesn't have focus (it depends on the browser+screen-reader combination).

**Comment by tugorez** [comment]
> Also, these announcements are polite -- so screen readers are totally allowed to elide some of them if they are coming in rapid succession.

Great catch. The original internal bug is related to the Form widget which uses the `assertive` mode tho. Just updated the example.

I'm using orca and I'm still able to reproduce the issue. By clicking multiple times I realize some announcements are still dropped.

**Comment by math1man** [comment]
I'm reproducing the issue with VoiceOver on a Mac. Changing the assertiveness doesn't change the behavior. I'm not surprised that it behaves differently on different screen reader services, but it should function correctly for all major services.

**Comment by marcianx** [comment]
[Here's a Web native demo][1] using the exact strategy that Flutter `accessibility.dart` uses for announcements. It works perfectly well for me on ChromeVox. Are you all able to reproduce the issue on this demo with your screen readers? This is a very standard approach.

[1]: https://jsbin.com/rufabelale/edit?html,output

**Comment by math1man** [comment]
I can reproduce this exact same issue using your demo on my Mac using VoiceOver, which doesn't surprise me at all. It seems like some semi-hacky solution like the proposal by @tugorez to suffix a period might be warranted.

**Comment by tugorez** [comment]
On a side note and just out of curiosity: @marcianx why is removing the previously announced messages needed?

**Comment by marcianx** [comment]
@math1man I'd like to clarify what you're seeing. Are you seeing the following:
1. You click on "Same announcements". You see the message pop-up and disappear quickly.
2. Wait for it to finish vocalizing completely (more realistic scenario).
3. You click on "Same announcements" again. You see the message pop-up and disappear, but you don't hear anything.

If so, that's very much a VoiceOver bug as it's not spec compliant.

Could you possibly test the behavior with the following modifications instead:
1. Changing it to "assertive" instead of polite
2. Increasing the duration of time before the message is removed
3. In the worst case, you could suffix with spaces. I don't know whether period could change the meaning of something.

@tugorez We remove the prior content to avoid the screen reader navigating into it at some later point since that might be confusing to users -- the message would likely no longer be relevant.

**Comment by tugorez** [comment]
@math1man does this solve the voiceover issue?
@marcianx does this WAI in chromevox?

```dart
import 'package:flutter/semantics.dart';
import 'package:flutter/material.dart';

const Color darkBlue = Color.fromARGB(255, 18, 32, 47);

var count = 0;
var addDot = false;

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      theme: ThemeData.dark().copyWith(
        scaffoldBackgroundColor: darkBlue,
      ),
      debugShowCheckedModeBanner: false,
      home: Scaffold(
        body: Center(
          child: MyWidget(),
        ),
      ),
    );
  }
}

class MyWidget extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        OutlinedButton(
          onPressed: () async {
            final textDirection = Directionality.of(context);
            await SemanticsService.announce('Test announcement $count', textDirection, assertiveness: Assertiveness.assertive);
            count++;
          },
          child: Text(
            'Different announcements',
            style: Theme.of(context).textTheme.headlineMedium,
          ),
        ),
        OutlinedButton(
          onPressed: () async {
            final textDirection = Directionality.of(context);
            final text = addDot ? 'Test announcement.' : 'Test announcement';
            addDot = !addDot;
            await SemanticsService.announce(text, textDirection,  assertiveness: Assertiveness.assertive);
          },
          child: Text(
            'Same announcements',
            style: Theme.of(context).textTheme.headlineMedium,
          ),
        ),
      ],
    );
  }
}
```

**Comment by math1man** [comment]
@marcianx yes I observe exactly the behavior you described.

> Changing it to "assertive" instead of polite

Flutter's Form widget uses `assertive`, it doesn't make a difference.

> Increasing the duration of time before the message is removed

This is deep inside the Flutter framework. I don't currently have the bandwidth to figure out how to change this, but I doubt it will matter.

**Comment by marcianx** [comment]
@math1man Sorry, I meant trying out these cases in the [JS Bin I linked upthread](https://jsbin.com/rufabelale/edit?html,output), which, I hope is straightforward to play around with. Fixing it within Flutter is straightforward for me, but it would really help to know what works on Web native before putting together a fix.

**Comment by math1man** [comment]
Ah I see, gotcha. I tried changing the duration to 1000 (1 second) and didn't observe any changes as long as I waited for the element to be removed. However, if I turn the delay way up and did not wait for the element to be removed, it would correctly announce each button press (albeit accummulating new elements).

**Comment by marcianx** [comment]
@math1man [Here is an alternative](https://jsbin.com/cajagejewa/edit?html,output) that appends space instead of period as a hack. Does this work for you?

**Comment by math1man** [comment]
As mentioned a few comments back, the space character doesn't work, but a non-breaking space `\u00A0` does work.
